### PR TITLE
Improve version

### PR DESCRIPTION
**Pull request by eternoendless**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Improve Version significantly in order for future-proofing. Read below.
| Type?         | improvement
| Category?     | CO
| BC breaks?    | A little (read below)
| Deprecations? | no
| Fixed ticket? | Maybe?
| How to test?  | Unit tested, should not have too big of an impact

# What's new

- Versions "1.7.5.0" and "7.5.0" are now considered equal. This effectively means that "1.7" is equal to "7" (instead of "17") regarding major version.
- It is now possible to have versions like "1.7.6.0-dev+nightly.20190131"
- `Version::buildFromString()` supports both legacy style versions (1.7.6.0) and semver style versions (7.6.0).
	- If you want the legacy version, you call `$version->getVersion()` (no BC here)
	- If you want the semver version, you call `$version->getSemVersion()`

Note: The constructor looks kind of wonky now, but I had to keep it like that in order to avoid breaking changes. I suggest using the static factory cited above only.

# What's changed (BC)

- `AppKernel::MAJOR_VERSION` is now "7" instead of "17".
- "1.2.4" is no longer greater than "1.2.3.4" (the latter is now interpreted as "2.3.4") 🤯

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/12251)

### CODE REVIEW

File: src/Core/Foundation/Version.php [hunk 1]
```diff
@@ -132,14 +212,51 @@ public function getMinorVersion()
...
      *
+     * @deprecated Since 1.7.6.0, use getPatchVersion();
+     *
      * @return int
      */
     public function getReleaseVersion()
     {
-        return $this->releaseVersion;
+        @trigger_error(
+            'getReleaseVersion() is deprecated since version 1.7.6.0 Use getPatchVersion() instead.',
```
Comment (PierreRambaud) [inline src/Core/Foundation/Version.php:235]: since 7.6.0 instead?

File: tests/Unit/Core/Foundation/VersionTest.php [hunk 2]
```diff
@@ -0,0 +1,624 @@
...
+                    'minor' => 1,
+                    'patch' => 2,
+                    'preRelease' => '',
+                    'buildMeta' => '',
+                ],
+            ],
+        ];
+    }
+
+    public function getCompareGreater()
```
Comment (PierreRambaud) [inline tests/Unit/Core/Foundation/VersionTest.php:529]: Can you add tests for `compare` to make sure `version_compare` is working properly with `.7.7.0-dev+nightly.20190526`, `1.7.7.0-beta.1+build.156`, `1.7.6.0+test.build` for example.

**Review by PierreRambaud — CHANGES_REQUESTED** [review]
See comments

File: src/Core/Foundation/Version.php (hunk 1 above)
Comment (eternoendless) [reply src/Core/Foundation/Version.php:235]: Officially versions are still 1.x.x.x

**Comment by PierreRambaud** [comment]
Waiting for rebase

**Comment by eternoendless** [comment]
Rebased 👍

**Review by PierreRambaud — DISMISSED** [review]

**Review by PierreRambaud — APPROVED** [review]

**Comment by PierreRambaud** [comment]
Thanks @eternoendless

**Pull request merged**